Add PlayerPrefs tutorial. (#1)
realm/unity-examples: tutorials/persistance-comparison-series/PlayerPrefs/player_prefs.md

- Part 1: PlayerPrefs *(this tutorial)*
- Part 2: Files *(coming soon)*
- Part 3: BinaryReader and BinaryWriter
- Part 4: SQL
- Part 5: Realm Unity SDK
- Part 6: Comparison of all those options

To make it easier to follow along we have prepared an example repository for you. All those examples can be found within the same Unity project since they all use the same example game so you can see the differences between those persistence approaches better.

The repository can be found at https://github.com/realm/unity-examples with this tutorial being on the [persistence-comparison](https://github.com/realm/unity-examples/tree/persistence-comparison) branch next to other tutorials we have prepared for you.

## Example game

*Note that if you have worked through any of the other tutorials in this series you can skip this section since we are using the same example for all parts of the series so that it is easier to see the differences between the approaches.*

The goal of this tutorial series is to show you a quick and easy way to make some first steps in the various ways to persist data in your game.

> Therefore the example we will be using will be as simply as possible in the editor itself so that we can fully focus on the actual code we need to write.

Reviewer: as simple as

Author: f0e7a1f

<img src="images/02_add_a_capsule.jpg" alt="Capsule in scene" width="500"/>

When you open up a clean 3D template, all you need to do is choosing `GameObject` -> `3D Object` -> `Capsule`.

You can then add scripts to the capsule by activating it in the hierarchy and using `Add Component` in the inspector.

<img src="images/03_add_script.jpg" alt="Capsule in scene" width="250"/>
<img src="images/04_script_name.jpg" alt="Capsule in scene" width="250"/>

The scripts we will add to this capsule showcasing the different methods will all have the same basic structure that can be found in `HitCountExample.cs`.

```cs
using UnityEngine;

/// <summary>
/// This script shows the basic structure of all other scripts.
/// </summary>
public class HitCountExample : MonoBehaviour
{
    // Keep count of the clicks.
    [SerializeField] private int hitCount; // 1

    private void Start() // 2
    {
        // Read the persisted data and set the initial hit count.
        hitCount = 0; // 3
    }

    private void OnMouseDown() // 4
    {
        // Increment the hit count on each click and save the data.
        hitCount++; // 5
    }
}

```

> The first thing we need to add is a counter for the clicks on the capsule (1). Add a `[SerilizeField]` here so that you can observe it while clicking on the capsule in the Unity editor.

Reviewer: SerializeField

Author: d373c55

    private readonly string hitCountKey = "HitCountKey"; // 1

    private void Start()
    {
        // Check if the key exists. If not, we never saved the hit count before.
        if (PlayerPrefs.HasKey(hitCountKey)) // 2
        {
            // Read the hit count from the PlayerPrefs.
            hitCount = PlayerPrefs.GetInt(hitCountKey); // 3
        }
    }

    private void OnMouseDown()
    {
        hitCount++;

        // Set and save the hit count before ending the game.
        PlayerPrefs.SetInt(hitCountKey, hitCount); // 4
        PlayerPrefs.Save(); // 5
    }

}

```

For the `PlayerPrefs` example we create a script named `PlayerPrefsExampleSimple` based on the `HitCountExample` shown earlier.

In addition to the basic structure we also need to define a key (1) that will be used to save the `hitCount` in the `PlayerPrefs`, let's call it `"HitCountKey"`.

When the game starts we first want to check if there was already a hit count saved. The `PlayerPrefs` have a built-in function `HasKey(hitCountKey)` (2) that let's us achieve exactly this. If the key exists we read it using `GetInt(hitCountKey)` (3) and save it in the counter.

> The second part is saving data whenever it changes. On each click after we incremented the `hitCount` we have to call `SetInt(key, value)` on `PlayerPrefs` (4) to set the new data. Note that this does not save the data to disk. This only happens during `OnApplicationQuit()` implicitly. We can explicitly write the data to disk at any time to avoid loosing data in case the game crashes and `OnApplicationQuit()` never gets called.

Reviewer: losing data

Author: d373c55

Instead of just detecting a mouse click the extended script will detect `Shift+Click` and `Ctrl+Click` as well.

Again, to visualize this in the editor we will add some more `[SerializeFields]` (1). Substitute the current one (`hitCount`) with the following:

```cs
// 1
[SerializeField] private int hitCountUnmodified = 0;
[SerializeField] private int hitCountShift = 0;
[SerializeField] private int hitCountControl = 0;
```

Each type of click will be shown in its own `Inspector` element.

The same has to be done for the `PlayerPrefs` keys. Remove the `HitCountKey` and add three new elements (2).

```cs
// 2
private readonly string HitCountKeyUnmodified = "HitCountKeyUnmodified";
private readonly string HitCountKeyShift = "HitCountKeyShift";
private readonly string HitCountKeyControl = "HitCountKeyControl";
```

There are many different ways to save more complex data. Here we will be using three different entries in `PlayerPrefs` as a first step. Later we will also look at how we can save structured data that belongs together in a different way.

One more field we need to save is the `KeyCode` for the key that was pressed:

```cs

> private KeyCode keyPressed = default;

Reviewer 2: Should we call this `modifier`?

Author: 5edc3c4

The keys can be addressed via their name as string but the type safe way to do this is to use the class `KeyCode` which defines every key necessary. For our case this would be `KeyCode.LeftShift` and `KeyCode.LeftControl`.

Those checks use `Input.GetKey()` (7) and if one of the two was found, it will be saved as the `keyPressed` (8). If neither of them was pressed (9) we just reset `keyPressed` to the `default` (10) which we will use as a marker for an unmodified mouse click.

```cs
private void Update() // 6
{
    // Check if a key was pressed.
    if (Input.GetKey(KeyCode.LeftShift)) // 7
    {
        // Set the LeftShift key.
        keyPressed = KeyCode.LeftShift; // 8
    }
    else if (Input.GetKey(KeyCode.LeftControl)) // 7
    {
        // Set the LeftControl key.
        keyPressed = KeyCode.LeftControl; // 8
    }
    else // 9
    {
        // In any other case reset to default and consider it unmodified.
        keyPressed = default; // 10
    }
}
```

The same triplet can then also be found in the click detection:

```cs
private void OnMouseDown()
{
    // Check if a key was pressed.

>     if (keyPressed == KeyCode.LeftShift) // 11

Reviewer 2: Maybe this will be more readable as a switch?

Author: Agree!
d03481e

    }

    // Persist the data to disk.
    PlayerPrefs.Save(); // 16
}
```

First we check if one of those two was held down while the click happened (11) and if so, increment the corresponding hit counter (12). If not (13), the `unmodfied` counter has to be incremented (14).

Finally, we need to set each of those three counters individually (15) via `PlayerPrefs.SetInt()` using the three keys we defined earlier.

Like in the previous example, we also call `Save()` (16) at the end to make sure, data does not get lost if the game does not end normally.

When switching back to the Unity editor, the script on the capsule should now look like this:

<img src="images/05_extended_version.jpg" alt="Extended example" width="500"/>

## More complex data

(see `PlayerPrefsExampleJson.cs` in the repository for the finished version)

In the previous two sections we have seen how to handle two simple examples of persisting data in `PlayerPrefs`. What if they get more complex than that? What if you want to structure and group data together?

One possible approach would be to use the fact that `PlayerPrefs` can hold a `string` and save a `JSON` in there.

> First we need to figure out how to actually transform our data into JSON. The .NET framework as well as the `UnityEngine` framework offer a JSON serialize and deserializer to do this job for us. Both behave very similar but we will use Unity's own [`JsonUtility`](https://docs.unity3d.com/ScriptReference/JsonUtility.html) which [performs better in Unity than other similar JSON solution](https://docs.unity3d.com/Manual/JSONSerialization.html).

Reviewer 2: ```suggestion
First we need to figure out how to actually transform our data into JSON. The .NET framework as well as the `UnityEngine` framework offer a JSON serializer and deserializer to do this job for us. Both behave very similar but we will use Unity's own [`JsonUtility`](https://docs.unity3d.com/ScriptReference/JsonUtility.html) which [performs better in Unity than other similar JSON solutions](https://docs.unity3d.com/Manual/JSONSerialization.html).
```

Reviewer 3: As a reminder, this is marked as resolved but the changes aren't present.
It's the same for some of the following comments.

> To transform data to JSON we first need to create a container object. This has [some restriction](https://docs.unity3d.com/ScriptReference/JsonUtility.ToJson.html):

Reviewer 2: ```suggestion
To transform data to JSON we first need to create a container object. This has [some restrictions](https://docs.unity3d.com/ScriptReference/JsonUtility.ToJson.html):
```

[SerializeField] private int hitCountShift = 0;
[SerializeField] private int hitCountControl = 0;
```

All those will eventually be saved into the same `PlayerPrefs` field which means we only need one key (3):

```cs
// 3
private readonly string hitCountKey = "HitCountKeyJson";
```

As before, the `keyPressed` will indicate which modifier was used:

```cs
// 4
private KeyCode keyPressed = default;
```

In `Start()` we then need to read the JSON. As before, we check if the `PlayerPrefs` key exists (5) and then read the data, this time using `GetString()` (as opposed to `GetInt()` before).

Transforming this JSON into the actual object is then done using `JsonUtility.FromJson()` (6) which takes the string as an argument. It's a generic function and we need to provide the information about which object this JSON is supposed to be representing, in this case `HitCount`.

If the JSON could be read and transformed successfully we can set the hit count fields (7) to their three values.

```cs
private void Start()
{
    // 5
    // Check if the key exists. If not, we never saved to it.
    if (PlayerPrefs.HasKey(HitCountKey))
    {
        // 6

>         string jsonString = PlayerPrefs.GetString(HitCountKey);

Reviewer 2: These can be `var`-s, though I don't know if Unity devs prefer the explicit type. I do find it a bit repetitive `string jsonString = ..GetString(...)` - it's fairly obvious it's a string 😆 I would probably write it as:

```csharp
var json = PlayerPrefs.GetString(...);
var hitCount = JsonUtility.FromJson<HitCount>(...);
```

but it's a personal preference, so fine to keep it as is.

Author: Not sure about Unity developers. In this case it's definitely me just being used to explicit types from other languages. I'll never get rid of that habit I assume. :D
578743f

        HitCount hitCount = JsonUtility.FromJson<HitCount>(jsonString);

        // 7
        if (hitCount != null)
        {
            hitCountUnmodified = hitCount.Unmodified;
            hitCountShift = hitCount.Shift;
            hitCountControl = hitCount.Control;
        }
    }
}
```

The detection for the key that was pressed is identical to the extended example since it does not involve loading or saving any data but is just a check for the key during `Update()`:

```cs
private void Update() // 8
{
    // Check if a key was pressed.
    if (Input.GetKey(KeyCode.LeftShift)) // 9
    {
        // Set the LeftShift key.
        keyPressed = KeyCode.LeftShift; // 10
    }
    else if (Input.GetKey(KeyCode.LeftControl)) // 9
    {
        // Set the LeftControl key.
        keyPressed = KeyCode.LeftControl; // 10
    }
    else // 11
    {
        // In any other case reset to default and consider it unmodified.
        keyPressed = default; // 12
    }
}
```

In a very similar fashion `OnMouseDown()` needs to save the data whenever it's changed.

```cs
private void OnMouseDown()
{

>     if (keyPressed == KeyCode.LeftShift) // 13

Reviewer 2: This would also look simpler as a switch

Author: Also included in d03481e.

    {
        // Increment the hit count and set it to PlayerPrefs.
        hitCountShift++; // 14
    }
    else if (keyPressed == KeyCode.LeftControl) // 13
    {
        // Increment the hit count and set it to PlayerPrefs.
        hitCountControl++; // 14
    }
    else // 15
    {
        // Increment the hit count and set it to PlayerPrefs.
        hitCountUnmodified++; // 16
    }

    // 17

>     HitCount hitCount = new();
>     hitCount.Unmodified = hitCountUnmodified;
>     hitCount.Shift = hitCountShift;
>     hitCount.Control = hitCountControl;

Reviewer 2: This can be simplified a little by using the property initializer syntax:

```suggestion
    var updatedCount = new HitCount
    { 
        Unmodified = hitCountUnmodified,
        Shift = hitCountShift,
        Control = hitCountControl,
    };
```

    // 18
    string jsonString = JsonUtility.ToJson(hitCount);
    PlayerPrefs.SetString(HitCountKey, jsonString);
    PlayerPrefs.Save();
}
```

Compared to before you see that checking the key and increasing the counter (13 - 16) is basically unchanged except for the save part that is now a bit different.

First, we need to create a new `HitCount` object (17) and assign the three counts. Using `JsonUtility.ToJson()` we can then (18) create a JSON string from this object and set it using the `PlayerPrefs`.

Remember to also call `Save()` here to make sure data cannot get lost in case the game crashes without being able to call `OnApplicationQuit()`.

Run the game and after you clicked the capsule a couple of time with or without Shift and Control, have a look at the result. The following screenshot shows the Windows registry which is where the `PlayerPrefs` get saved.

> The location is `HKEY_CURRENT_USER\SOFTWARE\Unity\UnityEditor\MongoDB Inc.\UnityPersistenceExample` and as you can see our JSON is right there, saved in plain text. Which is also one of the big down sides to keep in mind when using `PlayerPrefs`: data is not save and can easily be edited.

Reviewer 2: That location is probably not the same for all users, so probably worth replacing `MongoDB Inc.` with `*your-company*` or something like that

Reviewer 2: Also:
> data is not save

should probably be `data is not safe`?

Author: > That location is probably not the same for all users, so probably worth replacing `MongoDB Inc.` with `*your-company*` or something like that

It's the location using our example. But I should probably clarify that further here.

Author: 26fb8c0
